src/components/fader.js

```diff
@@ -0,0 +1,69 @@
+/**
+ * Creates a box around the element (assumed to be the camera's PoV) which can be used for fade-to-black.
+ */
+
+const FADE_DURATION_MS = 750;
+
+AFRAME.registerComponent("fader", {
+  schema: {
+    direction: { type: "string", default: "none" } // "in", "out", or "none"
+  },
+
+  init() {
+    const mesh = new THREE.Mesh(
+      new THREE.BoxGeometry(),
+      new THREE.MeshBasicMaterial({ color: 0x0, side: THREE.BackSide, opacity: 0, transparent: true })
+    );
+    mesh.scale.x = mesh.scale.y = mesh.scale.z = 0.05;
+    mesh.matrixNeedsUpdate = true;
+    this.el.object3D.add(mesh);
+    this.mesh = mesh;
+  },
+
+  async fadeOut() {
+    this.el.setAttribute("fader", { direction: "out" });
+
+    return new Promise(res => {
+      if (this.mesh.material.opacity === 1) {
+        res();
+      } else {
+        this._resolveFinish = res;
```

Comment:
I can already see the bug happening where you call fade in/out before a previous promise has resolved and having the previous one never resolve... Maybe keep an array of pending resolves? or immediately reject a previous one if it is interrupted...

Reply:
ah yeah, i was going to do the array thing which seemed overly complex, but i like the idea of just rejecting if you try to overwrite the existing promise.